pkg/datastore/filedb/filter.go

```diff
@@ -15,19 +15,181 @@
 package filedb
 
 import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+
 	"github.com/pipe-cd/pipecd/pkg/datastore"
 )
 
-// TODO: Implement filterable interface for each collection.
 type filterable interface {
 	Match(e interface{}, filters []datastore.ListFilter) (bool, error)
 }
 
 func filter(col datastore.Collection, e interface{}, filters []datastore.ListFilter) (bool, error) {
+	// If the collection implement filterable interface, use it.
 	fcol, ok := col.(filterable)
-	if !ok {
+	if ok {
+		return fcol.Match(e, filters)
+	}
+
+	// remarshal entity as map[string]interface{} struct.
+	raw, err := json.Marshal(e)
+	if err != nil {
+		return false, err
+	}
+	var omap map[string]interface{}
+	if err := json.Unmarshal(raw, &omap); err != nil {
+		return false, err
+	}
```

Comment:
Any way to deal with the entity struct directly instead of remarshal like this?
I feel that we are having many marshal-unmarshal around this part.

Reply:
To iterate throw struct and get field name as string, we only have using reflect or remarshal it to `map[string]interface{}` for now. I tried to do filter before first unmarshal (from object parts to object full data) but it does not work, since in some cases, the fields' values are stored duplicated between object parts, which make it's required to merge (unmarshal all object parts and build the whole object) once, then filter based on the fulfilled object. 

Besides, I read around the JSON marshal logic of the standard lib, and looks like it's called good/fast enough compare with others marshaling packages. So for now, I guess, I would like to make a try this method. If it's too costly this way, I will reconsider using reflect package or such. Wdyt?

Reply:
Totally agree about that. Let's give it a try and see how it work. 👍

Reply:
Thanks 🙌